### ISSUE TITLE
Change `XAxis.ticks` type from `(string | number)[]` to `any[]` if `XAxis.tickFormatter` is overwritten

### ISSUE THREAD

**Issue by GregoirePelegrin** [recharts/recharts#3678]
- [X] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
This feature allows for more flexibility in ticks management.
When handling dates, in order to make the chart more readable, the user may want a differentiable handling of the ticks values. This can be made easier by authorizing `XAxis.ticks` to be `any[]` instead of `(string | number)[]`.
This can be less clear for a user that `string` and `number` are natively supported, but can provide more flexibility.

### What does the proposed API look like?
While I am not clear on the details, this would imply changing the `XAxis` interface definition to implement the changed type, as well as any other place the type is assumed to be `(string | number)[]`. This could also return an error if `XAxis.ticks` type is different from `(string | number)[]` and `XAxis.tickFormatter` has not been overwritten.

**Comment by ckifer** [comment]
Hi!

Decreasing type safety isn't really a great path forward here, especially when ticks needs to be a string or number at the end of the day (for processing purposes). We could add to the union type here if there is a specific type you would like to support (a Date for example) but using any is a step backwards and would cause a lot more problems (now and down the line) than it is worth. 

Can you give an example of how you would use this? From there we can potentially see how the type (or something else) needs to change

**Comment by GregoirePelegrin** [comment]
I currently have an issue with the default handling of dates as x values. The goal is to have meaningful ticks (first of each month, every fortnight, ...). This wasn't the case as far as I could tell (but maybe have I looked wrong?), so I resorted to do the computation on my end, no big deal.

Here I am, with the timestamps of my meaningful ticks. I now want to display these timestamps such that it will still be readable, i.e. not display the year at every tick, but only on the first tick, last tick, and any tick that would correspond to the "01/01"s in between.

The intermediate ticks can be handled through `XAxis.tickFormatter` (checking if it is a "01/01" and returning the adequate format). This way will not permit me to check if this is the first or last tick though. In the case of no "01/01" in the ticks list, this would mean no year indication at all on the x axis.

Another way around, which would require decreasing type safety, would be to pass an object extending an interface such as the following: 
```
interface TickParentDefinition {
    value: (string | number)
}
```
This way, operators are already defined on the `value` attribute, but users can add any attribute of any kind they want, which they can then use in the `XAxis.tickFormatter` method (or any other).
This would then be available: 
```
interface MyTick extends TickParentDefinition {
    year: boolean
}

const myTickFormatter = (tickItem: MyTick) => {
    const result: string = moment(tickItem.value).format("DD/MM");
    if (tickItem.year || result === "01/01") {
        return moment(tickItem.value).format("DD/MM/YYYY");
    }
    return result;
};
```

If anything here isn't coherent with already existing code, please feel free to point it out, I admit I am not a recharts expert!

**Comment by ckifer** [comment]
Does this example kind of represent what you want to do? 
https://codesandbox.io/s/simple-line-chart-forked-9464mr

The second param to tickFormatter is the index of the item. You can check for first and last items that way

**Comment by GregoirePelegrin** [comment]
I wasn't aware of this possibility! While it allows me to check for `index === 0`, I cannot access the size of my list in this place. Thus, it only partially solves my issue.
It would solve my issue had I had access to the list from inside the `tickFormatter` method.
This also is a reason why I wanted to pass an object (with mandatory attributes) for the method to be able to access the relevant data stored inside this element.

**Comment by ckifer** [comment]
Hmm, I'm confused. If you pass the data to the chart, you have the list and you should be able to check the length of the list as I do in the above codesandbox without a direct reference to the data as a parameter to tickFormatter. Unless there is some issue when loading data async that I'm not thinking about it.

The proposed solution only works if conditional typing can be done correctly (if you provide a list of ticks yourself they are of type `T[]` (defaulted to `(string | number)[]` and unless you provide a `tickFormatter` that returns `string | number` your chart will be broken

**Comment by GregoirePelegrin** [comment]
This would be representative of my current situation: 
```
export const ChartComponent = (props: ChartComponentProps) => {
    const xTicksFormatter = (timestamp: number, index: number): string {
        ...
    };

    const histogramBuilder = () => {
        const data: {date: number, updates: number}[] = dataReceiver.obtainData();
        const xTicks: number[] = dateReceiver.getXTicks();
        const yTicks: number[] = dateReceiver.getYTicks();

        return (
            <Box display={"flex"} flexDirection={"column"} height={400}>
                <ResponsiveContainer width={"100%"} height={"100%"}>
                    <LineChart data={data} height={300} width={800}>
                        <XAxis dataKey={"date"} type={"number"} scale={"auto"} domain={[xTicks[0], xTicks[xTicks.length-1]]} ticks={xTicks} tickFormatter={xTicksFormatter}/>
                        <YAxis domain={[yTicks[0], yTicks[yTicks.length-1]]} ticks={yTicks}/>
                        <Line dataKey={"updates"} type={"monotone"} unit={"nbr"} dot={false} name={"Updates"}/>
                    </LineChart>
                </ResponsiveContainer>
            </Box>
        );
    };

    return (
        <>
            ...
            {histogramBuilder()}
        </>
    );
}
```

I cannot extract the data outside this method, and cannot store it in a state (not a technical limitation, but the rules I've been given here).

What I propose is to have ticks of type `TickParentDefinition[]` (or any interface `MyTickDefinition` extending `TickParentDefinition`), and have `tickFormatter(tickItem: MyTickDefinition, index: number): (string | number);` as `tickFormatter`, with the default being something like:
```
const tickFormatter = (tickItem: TickParentDefinition): (string | number) => {
    return tickItem.value;
}
```

**Comment by ckifer** [comment]
I guess I may sound like a broken record a bit, but you do have access to your data length there and you don't have to store it in state. You can move the formatter function into histogram builder underneath where you get the data, it makes no difference to the anatomy of your component as long as that is an accurate representation of your situation. 

There is no unseen rule here, I'm just trying to get you to a solution because I can't promise that this will be worked on in anything but responses to this issue any time soon which becomes a blocker for you (and I'd rather not have that if I can help it). I see your suggested approach and like it, but I'm not sure I'll get to it in a decent time frame. If you'd like to prepare a PR I'd be happy to review, else I'll probably keep trying to get it to work without changes 😅. Thanks!

**Comment by GregoirePelegrin** [comment]
> You can move the formatter function into histogram builder underneath where you get the data
You're right, this also solves my issue!

I must say I prefer my solution, for it doesn't feel like a workaround.
I understand if you do not have time for this as it is a minor issue which is not widespread, as well as introducing breaking changes for legacy.

I might try to work on a PR if I find it in me to explore the entirety of the code to try and implement my suggestion.

Thanks a lot for your time and understanding!

**Comment by ckifer** [comment]
> I must say I prefer my solution, for it doesn't feel like a workaround.

Makes sense, a workaround is what I was going for because I want to unblock you. Otherwise you may be waiting a while with current priorities. I don't have much time for recharts at the moment else I would happily try to get something changed.

No problem, you as well!

**Comment by GregoirePelegrin** [comment]
I must say that after trying to work on and off on this issue for the past week, I haven't been able to understand nor even locate the underlying mechanics for the basic handling of the ticks. 
Could someone tell me, in [this CodeSandbox example](https://codesandbox.io/s/sharp-sound-4737vh?file=/src/App.tsx) for instance, what is the sequence of methods called and used to handle the ticks? Or at least the most significant ones?
Thanks a lot!

**Comment by ckifer** [comment]
@GregoirePelegrin not sure what you're asking. Are you asking where in the source code is it handled?
